File: backend/database/models/task.py
from typing import Optional, List
from enum import Enum
from datetime import datetime
from sqlalchemy import String, Integer, DateTime, ForeignKey, Enum as SqlEnum, Boolean
from sqlalchemy.orm import mapped_column, relationship, Mapped

from .base import BaseModel
from .user import User
from .group import Group


class TaskStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ARCHIVED = "archived"
    CANCELLED = "cancelled"


class TaskPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class Task(BaseModel):
    __tablename__ = "tasks"

    title: Mapped[str] = mapped_column(String(255))
    description: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)

    priority: Mapped[TaskPriority] = mapped_column(
        SqlEnum(
            TaskPriority,
            name="task_priority",
            values_callable=lambda enum: [e.value for e in enum],
        ),
        nullable=False,
        default=TaskPriority.LOW,
    )
    status: Mapped[TaskStatus] = mapped_column(
        SqlEnum(
            TaskStatus,
            name="task_status",
            values_callable=lambda enum: [e.value for e in enum],
        ),
        nullable=False,
        default=TaskStatus.PENDING,
    )
    due_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    is_in_my_day: Mapped[bool] = mapped_column(Boolean, nullable=True, default=False)
    starred: Mapped[bool] = mapped_column(Boolean, nullable=True, default=False)

    estimated_pomodoros: Mapped[int] = mapped_column(Integer, nullable=True, default=0)
    completed_pomodoros: Mapped[int] = mapped_column(Integer, nullable=True, default=0)

    estimated_focus_time: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True, default=None
    )
    actual_focus_time: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True, default=None
    )

    complexity_level: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True, default=None
    )
    requires_deep_focus: Mapped[bool] = mapped_column(
        Boolean, nullable=True, default=False
    )
    optimal_session_length: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True, default=None
    )

    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"))
    group_id: Mapped[Optional[int]] = mapped_column(
        Integer, ForeignKey("groups.id"), nullable=True
    )

    user: Mapped["User"] = relationship("User", back_populates="tasks")
    group: Mapped["Group"] = relationship("Group", back_populates="tasks")
    tags: Mapped[List["TaskTag"]] = relationship("TaskTag", back_populates="task")  # type: ignore
    pomodoro_sessions: Mapped[List["PomodoroSession"]] = relationship(  # type: ignore
        "PomodoroSession", back_populates="task"
    )
    focus_sessions: Mapped[List["FocusSession"]] = relationship(  # type: ignore
        "FocusSession", back_populates="task"
    )
    subtasks: Mapped[List["Subtask"]] = relationship(  # type: ignore
        "Subtask", back_populates="task", cascade="all, delete-orphan"
    )

    @property
    def total_focus_time_hours(self) -> float:
        """Get total focus time in hours."""
        return self.actual_focus_time / 3600 if self.actual_focus_time else 0

    @property
    def pomodoro_completion_percentage(self) -> float:
        """Calculate pomodoro completion percentage."""
        if self.estimated_pomodoros and self.estimated_pomodoros > 0:
            return min(
                100.0, (self.completed_pomodoros / self.estimated_pomodoros) * 100
            )
        return 0.0

    @property
    def focus_time_completion_percentage(self) -> float:
        """Calculate focus time completion percentage."""
        if self.estimated_focus_time and self.estimated_focus_time > 0:
            return min(
                100.0, ((self.actual_focus_time or 0) / self.estimated_focus_time) * 100
            )
        return 0.0

    def __repr__(self) -> str:
        return f"<Task {self.title}>"

    def to_dict(self) -> dict:
        # Calculate subtask metadata
        subtask_count = len(self.subtasks) if self.subtasks else 0
        completed_subtasks = 0
        if hasattr(self, "subtasks") and self.subtasks:
            completed_subtasks = sum(
                1 for subtask in self.subtasks if subtask.is_completed
            )

        has_subtasks = subtask_count > 0

        # Serialize tags for json safety
        tags_data = []

        try:
            if hasattr(self, "tags") and self.tags:
                for task_tag in self.tags:
                    if hasattr(task_tag, "tag") and task_tag.tag:
                        tags_data.append(
                            {
                                "id": task_tag.tag.id,
                                "name": task_tag.tag.name,
                                "color": task_tag.tag.color,
                            }
                        )
        except Exception as e:
            tags_data = []

        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "priority": self.priority.value if self.priority else None,
            "status": self.status.value if self.status else None,
            "due_date": self.due_date.isoformat() if self.due_date else None,
            "completed_at": (
                self.completed_at.isoformat() if self.completed_at else None
            ),
            "is_in_my_day": self.is_in_my_day,
            "starred": self.starred,
            "user_id": self.user_id,
            "group_id": self.group_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            # Tags
            "tags": tags_data,
            # Subtask metadata
            "has_subtasks": has_subtasks,
            "subtask_count": subtask_count,
            "completed_subtask_count": completed_subtasks,
            "subtask_completion_percentage": (
                round((completed_subtasks / subtask_count) * 100)
                if subtask_count > 0
                else 0
            ),
            "estimated_focus_time": self.estimated_focus_time,
            "actual_focus_time": self.actual_focus_time,
            "total_focus_time_hours": self.total_focus_time_hours,
            "complexity_level": self.complexity_level,
            "requires_deep_focus": self.requires_deep_focus,
            "optimal_session_length": self.optimal_session_length,
            "pomodoro_completion_percentage": self.pomodoro_completion_percentage,
            "focus_time_completion_percentage": self.focus_time_completion_percentage,
        }

File: frontend/src/types/task.ts
export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent'
export type TaskStatus = 'pending' | 'in_progress' | 'completed' | 'archived' | 'cancelled'

export interface Task {
  id: number
  title: string
  description?: string
  priority: TaskPriority
  status: TaskStatus
  due_date?: string
  completed_at?: string
  created_at: string
  updated_at: string
  estimated_pomodoros: number
  completed_pomodoros: number
  is_in_my_day: boolean
  starred: boolean
  user_id: number
  group_id?: number
  tags?: Tag[]
  subtasks?: Subtask[]
  // Subtask metadata (from backend)
  has_subtasks: boolean
  subtask_count: number
  completed_subtask_count: number
  subtask_completion_percentage: number
}

export interface Tag {
  id: number
  name: string
  color?: string
  user_id: number
}

export interface Subtask {
  id: number
  title: string
  description?: string
  position: number
  task_id: number
  is_completed: boolean
  created_at: string
  updated_at: string
}

export interface CreateSubtaskRequest {
  title: string
  description?: string
  position: number
  task_id: number
}

export interface UpdateSubtaskRequest {
  title?: string
  description?: string
  position?: number
  is_completed?: boolean
}

export interface SubtaskListResponse {
  subtasks: Subtask[]
  count: number
  task_id: number
}

export interface SubtaskStats {
  total: number
  completed: number
  pending: number
  completion_percentage: number
  task_id: number
}

export interface SubtaskFilters {
  completed?: boolean
  task_id?: number
  search?: string
}

export interface TaskGroup {
  id: number
  name: string
  description?: string
  color?: string
  user_id: number
  created_at: string
  updated_at: string
}

export interface CreateTaskRequest {
  title: string
  description?: string
  priority?: TaskPriority
  due_date?: string
  estimated_pomodoros?: number
  is_in_my_day?: boolean
  starred?: boolean
  group_id?: number
  tag_ids?: number[]
  user_id: number
}

export interface UpdateTaskRequest {
  title?: string
  description?: string
  priority?: TaskPriority
  status?: TaskStatus
  due_date?: string
  estimated_pomodoros?: number
  completed_pomodoros?: number
  is_in_my_day?: boolean
  starred?: boolean
  group_id?: number
  tag_ids?: number[]
}

export interface TaskFilters {
  status?: string[]
  priority?: string[]
  group_id?: number
  search?: string
  due_date_from?: string
  due_date_to?: string
  completed?: boolean
  starred?: boolean
  overdue?: boolean
  is_in_my_day?: boolean
  tags?: string[]
}

export interface TaskResponse {
  tasks: Task[]
  pagination: {
    total_count: number
    page: number
    page_size: number
    total_pages: number
    has_next: boolean
    has_prev: boolean
  }
  filters_applied?: TaskFilters
  sort?: {
    sort_by: string
    sort_order: string
  }
}

export interface TaskListResponse {
  tasks: Task[]
  count: number
  query?: string
}

// Pomodoro-related types
export interface PomodoroSession {
  id: number
  task_id: number
  duration: number
  completed: boolean
  start_time: string
  end_time?: string
  session_type: 'work' | 'short_break' | 'long_break'
  user_id: number
}

export interface PomodoroStats {
  total_sessions: number
  completed_sessions: number
  total_time: number
  sessions_today: number
  time_today: number
  average_session_length: number
  streak_days: number
}

// Task context and state types
export interface TaskState {
  tasks: Task[]
  selectedTask: Task | null
  filters: TaskFilters
  loading: boolean
  error: string | null
  pagination: {
    page: number
    per_page: number
    total: number
    total_pages: number
  }
}

export interface TaskContextType extends TaskState {
  // Task operations
  createTask: (task: CreateTaskRequest) => Promise<Task>
  updateTask: (id: number, updates: UpdateTaskRequest) => Promise<Task>
  deleteTask: (id: number) => Promise<void>
  duplicateTask: (id: number) => Promise<Task>
  
  // Task status operations
  toggleTaskCompletion: (id: number) => Promise<Task>
  markTaskAsStarred: (id: number, starred: boolean) => Promise<Task>
  
  // Subtask operations
  addSubtask: (taskId: number, title: string) => Promise<Subtask>
  updateSubtask: (taskId: number, subtaskId: number, updates: Partial<Subtask>) => Promise<Subtask>
  deleteSubtask: (taskId: number, subtaskId: number) => Promise<void>
  toggleSubtaskCompletion: (taskId: number, subtaskId: number) => Promise<Subtask>
  
  // Selection and filtering
  selectTask: (task: Task | null) => void
  setFilters: (filters: Partial<TaskFilters>) => void
  clearFilters: () => void
  
  // Data fetching
  fetchTasks: (params?: Partial<TaskFilters>) => Promise<void>
  refetchTasks: () => Promise<void>
  
  // Bulk operations
  bulkUpdateTasks: (taskIds: number[], updates: UpdateTaskRequest) => Promise<void>
  bulkDeleteTasks: (taskIds: number[]) => Promise<void>
}

// Task sorting and view options
export interface TaskSortOption {
  field: 'title' | 'priority' | 'due_date' | 'created_at' | 'updated_at'
  direction: 'asc' | 'desc'
}

export interface TaskViewOptions {
  groupBy: 'none' | 'priority' | 'status' | 'group' | 'due_date'
  sortBy: TaskSortOption
  showCompleted: boolean
  showSubtasks: boolean
  viewType: 'list' | 'grid' | 'kanban'
}

// Task statistics
export interface TaskStats {
  total: number
  completed: number
  pending: number
  in_progress: number
  overdue: number
  due_today: number
  due_this_week: number
  by_priority: Record<TaskPriority, number>
  by_group: Record<string, number>
  completion_rate: number
  average_completion_time: number
}

// Task activity and history
export interface TaskActivity {
  id: number
  task_id: number
  action: 'created' | 'updated' | 'completed' | 'deleted' | 'commented'
  description: string
  user_id: number
  created_at: string
  metadata?: Record<string, unknown>
}

// Task templates
export interface TaskTemplate {
  id: number
  name: string
  title: string
  description?: string
  priority: TaskPriority
  estimated_pomodoros: number
  tags: Tag[]
  subtasks: Omit<Subtask, 'id' | 'task_id'>[]
  user_id: number
}

// Note: ApiError is imported from @/types/api